src/hooks/common/raw_input.rs

```diff
@@ -0,0 +1,176 @@
+// mouse click (VK_...) and the values in mouse_down are updated.
+fn handle_raw_mouse_input(io: &mut Io, raw_mouse: &RAWMOUSE_0_0) {
+    let button_flags = raw_mouse.usButtonFlags as u32;
+
+    let has_flag = |flag| button_flags & flag != 0;
+    let mut set_key_down = |VIRTUAL_KEY(index), val: bool| io.keys_down[index as usize] = val;
+
+    // Check whether any of the mouse buttons was pressed or released.
+    if has_flag(RI_MOUSE_LEFT_BUTTON_DOWN) {
+        set_key_down(VK_LBUTTON, true);
+        io.mouse_down[0] = true;
+    }
+    if has_flag(RI_MOUSE_LEFT_BUTTON_UP) {
+        set_key_down(VK_LBUTTON, false);
+        io.mouse_down[0] = false;
+    }
+    if has_flag(RI_MOUSE_RIGHT_BUTTON_DOWN) {
+        set_key_down(VK_RBUTTON, true);
+        io.mouse_down[1] = true;
+    }
+    if has_flag(RI_MOUSE_RIGHT_BUTTON_UP) {
+        set_key_down(VK_RBUTTON, false);
+        io.mouse_down[1] = false;
+    }
+    if has_flag(RI_MOUSE_MIDDLE_BUTTON_DOWN) {
+        set_key_down(VK_MBUTTON, true);
+        io.mouse_down[2] = true;
+    }
+    if has_flag(RI_MOUSE_MIDDLE_BUTTON_UP) {
+        set_key_down(VK_MBUTTON, false);
+        io.mouse_down[2] = false;
+    }
+    if has_flag(RI_MOUSE_BUTTON_4_DOWN) {
+        set_key_down(VK_XBUTTON1, true);
+        io.mouse_down[3] = true;
+    }
+    if has_flag(RI_MOUSE_BUTTON_4_UP) {
+        set_key_down(VK_XBUTTON1, false);
+        io.mouse_down[3] = false;
+    }
+    if has_flag(RI_MOUSE_BUTTON_5_DOWN) {
+        set_key_down(VK_XBUTTON2, true);
+        io.mouse_down[4] = true;
+    }
+    if has_flag(RI_MOUSE_BUTTON_5_UP) {
+        set_key_down(VK_XBUTTON2, false);
+        io.mouse_down[4] = false;
+    }
+
+    // Apply vertical mouse scroll.
+    if button_flags & RI_MOUSE_WHEEL != 0 {
+        let wheel_delta = raw_mouse.usButtonData as i16 / WHEEL_DELTA as i16;
+        io.mouse_wheel += wheel_delta as f32;
+    }
+
+    // Apply horizontal mouse scroll.
+    if button_flags & RI_MOUSE_HWHEEL != 0 {
+        let wheel_delta = raw_mouse.usButtonData as i16 / WHEEL_DELTA as i16;
+        io.mouse_wheel_h += wheel_delta as f32;
+    }
+}
+
+// Handle raw keyboard input.
+fn handle_raw_keyboard_input(io: &mut Io, raw_keyboard: &RAWKEYBOARD) {
+    // Ignore messages without a valid key code
+    if raw_keyboard.VKey == 0 {
+        return;
+    }
+
+    // Extract the keyboard flags.
+    let flags = raw_keyboard.Flags as u32;
+
+    // Compute the scan code, applying the prefix if it is present.
+    let scan_code = {
+        let mut code = raw_keyboard.MakeCode as u32;
+        // Necessary to check LEFT/RIGHT keys on CTRL & ALT & others (not shift)
+        if flags & RI_KEY_E0 != 0 {
+            code |= 0xe000;
+        }
+        if flags & RI_KEY_E1 != 0 {
+            code |= 0xe100;
+        }
+        code
+    };
+
+    // Check the key status.
+    let is_key_down = flags == RI_KEY_MAKE;
+    let is_key_up = flags & RI_KEY_BREAK != 0;
```

Comment:
In general, I want to start pushing this repo in the direction in which everything is organized and commented in this way. I've been slacking off on that, but I realized that this way the understandability of a complex codebase skyrockets.